### ISSUE TITLE
Hang on RequestContext.WriteBody()

### ISSUE THREAD

**Issue by Rurouni** [Cysharp/YetAnotherHttpHandler#94]
Hi,
We've run into an interesting bug: after some time (the repro is not clear), some requests hang forever.
Looking at call stacks during the hang there are 3 threads in deadlock:

1) sits in forever while loop after obtaining _handleLock
 ```
[Native Transition]
Thread.Sleep()
RequestContext.WriteBody()
RequestContext.<RunReadRequestLoopAsync>d__25.MoveNext()
AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext()
ExecutionContext.RunInternal() [2]
ExecutionContext.Run() [2]
AsyncMethodBuilderCore.MoveNextRunner.Run()
ValueTaskAwaiter.<>c.<.cctor>b__9_0()
ThreadPoolScheduler.<>c.<Schedule>b__0_0()
QueueUserWorkItemCallback.WaitCallback_Context()
ExecutionContext.RunInternal() [1]
ExecutionContext.Run() [1]
QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
ThreadPoolWorkQueue.Dispatch()
_ThreadPoolWaitCallback.PerformWaitCallback()
```
2) waits for _handleLock
```
RequestContext.Release() at \Library\PackageCache\com.cysharp.yetanotherhttphandler@0d7390aec2\RequestContext.cs:line 75
    NativeHttpHandlerCore.OnComplete() at \Library\PackageCache\com.cysharp.yetanotherhttphandler@0d7390aec2\NativeHttpHandlerCore.cs:line 476
```
3) waits for _handleLock
```
RequestContext.TryAbort() at \Library\PackageCache\com.cysharp.yetanotherhttphandler@0d7390aec2\RequestContext.cs:line 248
    ResponseContext.Cancel() at \Library\PackageCache\com.cysharp.yetanotherhttphandler@0d7390aec2\ResponseContext.cs:line 153
    ResponseContext.<>c.<.ctor>b__8_0()
    CancellationCallbackInfo.ExecutionContextCallback()
    ExecutionContext.RunInternal() [2]
    ExecutionContext.Run() [2]
    ExecutionContext.Run()
    CancellationCallbackInfo.ExecuteCallback()
    CancellationTokenSource.CancellationCallbackCoreWork()
    CancellationTokenSource.ExecuteCallbackHandlers()
    CancellationTokenSource.NotifyCancellation()
    CancellationTokenSource.Cancel()
    CancellationTokenSource.Cancel()
    GrpcCall<PlayerActionDto, Result<ExecutePlayerActionResponse>>.CancelCall()
    GrpcCall<PlayerActionDto, Result<ExecutePlayerActionResponse>>.Cleanup()
    GrpcCall<PlayerActionDto, Result<ExecutePlayerActionResponse>>.<RunCall>d__74.MoveNext()
    AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext()
    ExecutionContext.RunInternal() [1]
    ExecutionContext.Run() [1]
    AsyncMethodBuilderCore.MoveNextRunner.Run()
    AwaitTaskContinuation.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
    ThreadPoolWorkQueue.Dispatch()
    _ThreadPoolWaitCallback.PerformWaitCallback()
```
Version: 1.5.1

I would greatly appreciate any help
Thank you

**Comment by Rurouni** [comment]
I've reproduced it with YahaEventSource enabled and it looks like this
1)
```
Start sending the request body: Grpc.Net.Client.Internal.PushUnaryContent`2[[GetPlayerAggregateRequest],[Shared.Networking.Result`1[[GetPlayerAggregateResponse]]
Cysharp.Net.Http.YahaEventSource:Info (string) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/YahaEventSource.cs:28)
Cysharp.Net.Http.NativeHttpHandlerCore:SendAsync (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/NativeHttpHandlerCore.cs:178)
Cysharp.Net.Http.YetAnotherHttpHandler:SendAsync (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/YetAnotherHttpHandler.cs:169)
System.Net.Http.DelegatingHandler:SendAsync (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)
```
2)
```
[ReqSeq:2921:State:0x1718] Sending the request body: Length=19
Cysharp.Net.Http.YahaEventSource:Trace (string) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/YahaEventSource.cs:23)
Cysharp.Net.Http.RequestContext:WriteBody (System.Span`1<byte>) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/RequestContext.cs:137)
Cysharp.Net.Http.RequestContext/<RunReadRequestLoopAsync>d__25:MoveNext () (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/RequestContext.cs:106)
```
3)
```
[ReqSeq:2921:State:0x1718] Send buffer is full.
Cysharp.Net.Http.YahaEventSource:Trace (string) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/YahaEventSource.cs:23)
Cysharp.Net.Http.RequestContext:WriteBody (System.Span`1<byte>) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/RequestContext.cs:167)
Cysharp.Net.Http.RequestContext/<RunReadRequestLoopAsync>d__25:MoveNext () (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/RequestContext.cs:106)
```
4)
```
[ReqSeq:2921] Response completed with failure (client error (SendRequest): operation was canceled) (0x0)
Cysharp.Net.Http.YahaEventSource:Trace (string) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/YahaEventSource.cs:23)
Cysharp.Net.Http.ResponseContext:CompleteAsFailed (string,uint) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/ResponseContext.cs:115)
Cysharp.Net.Http.RequestContext:CompleteAsFailed (string,uint) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/RequestContext.cs:291)
Cysharp.Net.Http.NativeHttpHandlerCore:OnComplete (int,intptr,Cysharp.Net.Http.CompletionReason,uint) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/NativeHttpHandlerCore.cs:451)
```
5)
```
[ReqSeq:2921:State:0x1718] Send buffer is full.
Cysharp.Net.Http.YahaEventSource:Trace (string) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/YahaEventSource.cs:23)
Cysharp.Net.Http.RequestContext:WriteBody (System.Span`1<byte>) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/RequestContext.cs:167)
Cysharp.Net.Http.RequestContext/<RunReadRequestLoopAsync>d__25:MoveNext () (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/RequestContext.cs:106)
```
6)
```
[ReqSeq:2921] Response was cancelled
Cysharp.Net.Http.YahaEventSource:Trace (string) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/YahaEventSource.cs:23)
Cysharp.Net.Http.ResponseContext:Cancel () (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/ResponseContext.cs:149)
Cysharp.Net.Http.ResponseContext/<>c:<.ctor>b__8_0 (object) (at ./Library/PackageCache/com.cysharp.yetanotherhttphandler@ceb2f24f34/ResponseContext.cs:31)
```

after that it just keeps logging `[ReqSeq:2921:State:0x1718] Send buffer is full` forever

@mayuki I can add more logs locally if that will help with the investigation

**Comment by mayuki** [comment]
Thank you for the detailed information.
I think this problem has been fixed in the 1.5.4 release, so please try it.
